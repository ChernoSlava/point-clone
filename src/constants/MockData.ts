export const mockArticles = [
  {
    id: 1,
    title: { short: "Article 1" },
    description: { intro: "Description 1" },
    thumbnail: "thumbnail1.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 2,
    title: { short: "Article 2" },
    description: { intro: "Description 2" },
    thumbnail: "thumbnail2.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 3,
    title: { short: "Article 3" },
    description: { intro: "Description 3" },
    thumbnail: "thumbnail3.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 4,
    title: { short: "Article 4" },
    description: { intro: "Description 4" },
    thumbnail: "thumbnail4.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 5,
    title: { short: "Article 5" },
    description: { intro: "Description 5" },
    thumbnail: "thumbnail5.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 6,
    title: { short: "Article 6" },
    description: { intro: "Description 6" },
    thumbnail: "thumbnail6.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 7,
    title: { short: "Article 7" },
    description: { intro: "Description 7" },
    thumbnail: "thumbnail7.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 8,
    title: { short: "Article 8" },
    description: { intro: "Description 8" },
    thumbnail: "thumbnail8.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 9,
    title: { short: "Article 9" },
    description: { intro: "Description 9" },
    thumbnail: "thumbnail9.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 10,
    title: { short: "Article 10" },
    description: { intro: "Description 10" },
    thumbnail: "thumbnail10.jpg",
    dates: { posted: "1630860959" },
  },
];

export const mockArticlesSmall = [
  {
    id: 1,
    title: { short: "Article 1" },
    description: { intro: "Description 1" },
    thumbnail: "thumbnail1.jpg",
    dates: { posted: "1630860959" },
  },
  {
    id: 2,
    title: { short: "Article 2" },
    description: { intro: "Description 2" },
    thumbnail: "thumbnail2.jpg",
    dates: { posted: "1630860959" },
  },
];
