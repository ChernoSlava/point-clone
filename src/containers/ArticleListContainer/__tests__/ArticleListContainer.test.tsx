import React from "react";
import { MemoryRouter } from "react-router-dom";
import { mockArticlesSmall } from "@constants/MockData";
import * as useArticleListModule from "@hooks/index";
import { render, screen } from "@testing-library/react";

import "@testing-library/jest-dom";

import { ArticleListContainer } from "..";

jest.mock("@hooks/index");

describe("ArticleListContainer tests", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders loading message while fetching data", () => {
    const mockUseArticleList = jest.fn();
    jest
      .spyOn(useArticleListModule, "useArticleList")
      .mockImplementation(mockUseArticleList);

    mockUseArticleList.mockReturnValue({
      loading: true,
      error: undefined,
      articles: [],
      loadingMore: false,
    });

    render(<ArticleListContainer />);

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  test("renders error message while fetching data", () => {
    const mockUseArticleList = jest.fn();
    jest
      .spyOn(useArticleListModule, "useArticleList")
      .mockImplementation(mockUseArticleList);

    mockUseArticleList.mockReturnValue({
      loading: false,
      error: new Error("Test error"),
      articles: [],
      loadingMore: false,
    });

    render(<ArticleListContainer />);

    expect(screen.getByText("Error: Test error")).toBeInTheDocument();
  });

  test("renders article list with correct data", () => {
    const mockUseArticleList = jest.fn();
    jest
      .spyOn(useArticleListModule, "useArticleList")
      .mockImplementation(mockUseArticleList);

    mockUseArticleList.mockReturnValue({
      loading: false,
      error: undefined,
      articles: mockArticlesSmall,
      loadingMore: false,
    });

    const { container, unmount } = render(
      <MemoryRouter>
        <ArticleListContainer />
      </MemoryRouter>
    );

    const imgElements = screen.queryAllByRole("img", { name: "Cartoon" });
    expect(imgElements.length).toBeGreaterThan(0);

    const firstImgElement = imgElements[0];
    expect(firstImgElement).toHaveAttribute(
      "src",
      "https://i.simpalsmedia.com/point.md/news/900x900/thumbnail1.jpg"
    );

    const thumbnailElements = screen.queryAllByRole("img", { name: "icon" });
    expect(thumbnailElements.length).toBeGreaterThan(0);

    const secondThumbnailElement = thumbnailElements[1];
    expect(secondThumbnailElement).toHaveAttribute(
      "src",
      "https://i.simpalsmedia.com/point.md/news/370x194/thumbnail2.jpg"
    );

    expect(screen.getByText("Article 1")).toBeInTheDocument();
    expect(screen.getByText("Description 1")).toBeInTheDocument();

    expect(screen.getByText("Article 2")).toBeInTheDocument();
    expect(screen.getByText("Description 2")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
    unmount();
  });

  test("renders loadingMore", () => {
    const mockUseArticleList = jest.fn();
    jest
      .spyOn(useArticleListModule, "useArticleList")
      .mockImplementation(mockUseArticleList);

    mockUseArticleList.mockReturnValue({
      loading: false,
      error: undefined,
      articles: mockArticlesSmall,
      loadingMore: true,
    });

    const { unmount } = render(
      <MemoryRouter>
        <ArticleListContainer />
      </MemoryRouter>
    );

    const loadingMoreElement = screen.getByText("Loading more...");
    expect(loadingMoreElement).toBeInTheDocument();
    unmount();
  });
});
