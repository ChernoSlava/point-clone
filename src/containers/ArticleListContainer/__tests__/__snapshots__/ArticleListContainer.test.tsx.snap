// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleListContainer tests renders article list with correct data 1`] = `
<div>
  <div
    class="sc-fsYfdN fGPEJW"
  >
    <h2
      class="sc-qZrbh dvdZjT"
    >
      Сегодня
    </h2>
    <ul
      class="sc-jsEeTM jidTqx"
    >
      <a
        href="/article/article-1"
      >
        <li
          class="sc-beySPh jOqDaV"
        >
          <div
            class="sc-guDLey dSXgup"
          >
            <img
              alt="Cartoon"
              class="sc-dmyCSP gctnZZ"
              src="https://i.simpalsmedia.com/point.md/news/900x900/thumbnail1.jpg"
            />
          </div>
          <div
            class="sc-hLQSwg ezfliH"
          >
            <h3
              class="sc-eDLKkx bCcaMp"
            >
              Article 1
            </h3>
            <p
              class="sc-jTQCzO gawhwJ"
            >
              Description 1
            </p>
            <p
              class="sc-iBdnpw giGobE"
            >
              <img
                alt="icon"
                class="sc-gLLuof gNHDaF"
                src="https://i.simpalsmedia.com/point.md/news/370x194/thumbnail1.jpg"
              />
              886 дней назад
            </p>
          </div>
        </li>
      </a>
      <a
        href="/article/article-2"
      >
        <li
          class="sc-beySPh jOqDaV"
        >
          <div
            class="sc-guDLey dSXgup"
          >
            <img
              alt="Cartoon"
              class="sc-dmyCSP gctnZZ"
              src="https://i.simpalsmedia.com/point.md/news/900x900/thumbnail2.jpg"
            />
          </div>
          <div
            class="sc-hLQSwg ezfliH"
          >
            <h3
              class="sc-eDLKkx bCcaMp"
            >
              Article 2
            </h3>
            <p
              class="sc-jTQCzO gawhwJ"
            >
              Description 2
            </p>
            <p
              class="sc-iBdnpw giGobE"
            >
              <img
                alt="icon"
                class="sc-gLLuof gNHDaF"
                src="https://i.simpalsmedia.com/point.md/news/370x194/thumbnail2.jpg"
              />
              886 дней назад
            </p>
          </div>
        </li>
      </a>
    </ul>
  </div>
</div>
`;
